File: src/main/java/io/bloco/faker/components/Time.java
package io.bloco.faker.components;

import org.joda.time.DateTime;

import java.util.Date;

import io.bloco.faker.FakerComponent;
import io.bloco.faker.FakerData;
import io.bloco.faker.helpers.Period;

public class Time extends FakerComponent {

    public Time(FakerData fakerData) {
        super(fakerData);
    }

    public Date between(Date from, Date to, Period period) {
        return between(new DateTime(from), new DateTime(to), period).toDate();
    }

    public Date between(Date from, Date to) {
        return between(from, to, Period.all);
    }

    public Date forward(int numberOfDays, Period period) {
        return randomTime(now(), period).plusDays(numberOfDays).toDate();
    }

    public Date forward(int numberOfDays) {
        return forward(numberOfDays, Period.all);
    }

    public Date backward(int numberOfDays, Period period) {
        return randomTime(now(), period).minusDays(numberOfDays).toDate();
    }

    public Date backward(int numberOfDays) {
        return backward(numberOfDays, Period.all);
    }

    private DateTime now() {
        return DateTime.now();
    }

    private DateTime between(DateTime from, DateTime to, Period period) {
        return randomTime(new DateTime(getRandomTimeBetweenTwoDates(from, to)), period);
    }

    private long getRandomTimeBetweenTwoDates(DateTime from, DateTime to) {
        long diff = to.getMillis() - from.getMillis() + 1;
        return from.getMillis() + (long) (Math.random() * diff);
    }

    private DateTime randomTime(DateTime dateTime, Period period) {
        return dateTime.withHourOfDay(hours(period))
                .withMinuteOfHour(minutes())
                .withSecondOfMinute(seconds());
    }

    private int hours(Period period) {
        int[] values = period.getValues();
        return values[randomHelper.number(values.length)];
    }

    private int minutes() {
        return randomHelper.number(60);
    }

    private int seconds() {
        return randomHelper.number(60);
    }
}

File: lib/src/main/kotlin/io/bloco/faker/components/Time.kt
package io.bloco.faker.components

import io.bloco.faker.FakerComponent
import io.bloco.faker.FakerData
import io.bloco.faker.helpers.Period
import io.bloco.faker.helpers.toDate
import io.bloco.faker.helpers.toLocalDateTime
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.temporal.ChronoField
import java.util.Date

class Time(fakerData: FakerData) : FakerComponent(fakerData) {
    fun between(from: Date, to: Date, period: Period): Date {
        return between(from.toLocalDateTime(), to.toLocalDateTime(), period)
    }

    fun between(from: Date, to: Date): Date {
        return between(from, to, Period.All)
    }

    @JvmOverloads
    fun forward(numberOfDays: Int, period: Period = Period.All): Date {
        return randomTime(now(), period).plusDays(numberOfDays.toLong()).toDate()
    }

    @JvmOverloads
    fun backward(numberOfDays: Int, period: Period = Period.All): Date {
        return randomTime(now(), period).minusDays(numberOfDays.toLong()).toDate()
    }

    private fun now(): LocalDateTime {
        return LocalDateTime.now()
    }

    private fun between(from: LocalDateTime, to: LocalDateTime, period: Period): Date {
        return randomTime(getRandomTimeBetweenTwoDates(from, to).toLocalDateTime(), period).toDate()
    }

    private fun getRandomTimeBetweenTwoDates(from: LocalDateTime, to: LocalDateTime): Long {
        val toInMillis = to.toEpochSecond(ZoneOffset.UTC) * 1000 + to.get(ChronoField.MILLI_OF_SECOND)
        val fromInMillis = from.toEpochSecond(ZoneOffset.UTC) * 1000 + from.get(ChronoField.MILLI_OF_SECOND)
        val diff = toInMillis - fromInMillis + 1
        return fromInMillis + (Math.random() * diff).toLong()
    }

    private fun randomTime(dateTime: LocalDateTime, period: Period): LocalDateTime {
        return dateTime.withHour(hours(period))
            .withMinute(minutes())
            .withSecond(seconds())
    }

    private fun hours(period: Period): Int {
        val values: IntArray = period.values
        return values[randomHelper.number(values.size)]
    }

    private fun minutes(): Int {
        return randomHelper.number(60)
    }

    private fun seconds(): Int {
        return randomHelper.number(60)
    }
}
